cpp/include/raft/spatial/knn/detail/ann_ivf_flat.cuh

```diff
@@ -0,0 +1,771 @@
+  rmm::device_uvector<T> list_data_dev_;
+  // The device memory pointer; inverted list for index; size [ninterleave_]
+  rmm::device_uvector<uint32_t> list_index_dev_;
+  // The device memory pointer; Used for list_data_manage_ptr_; size [nlist_]
+  rmm::device_uvector<uint32_t> list_prefix_interleaved_dev_;
+  // The device memory pointer; the number of each cluster(list); size [nlist_]
+  rmm::device_uvector<uint32_t> list_lengths_dev_;
+  // The device memory pointer; centriod; size [nlist_, dim_]
+  rmm::device_uvector<float> centriod_dev_;
+  // The device memory pointer; centriod norm ; size [nlist_, dim_]
+  rmm::device_uvector<float> centriod_norm_dev_;
+  // Memory pool for use during search; after the first search is done the pool is not likely to
+  // resize, saving the costs of allocations.
+  std::optional<rmm::mr::pool_memory_resource<rmm::mr::device_memory_resource>> search_mem_res;
+
+  // host pointer
+  //  The host memory pointer; inverted list for data; size [ninterleave_, dim_]
+  std::vector<T> list_data_host_;
+  // The host memory pointer; inverted list for index; size [ninterleave_]
+  std::vector<uint32_t> list_index_host_;
+  // The host memory pointer; Used for list_data_manage_ptr_; size [nlist_]
+  std::vector<uint32_t> list_prefix_interleaved_host_;
+  // The host memory pointer; the number of each cluster(list); size [nlist_]
+  std::vector<uint32_t> list_lengths_host_;
+
+  cuivflStatus_t cuivflBuildOptimizedKmeans(float* centriod_managed_ptr,
+                                            const T* dataset,
+                                            T* trainset,
+                                            uint32_t* clusterSize,
+                                            uint32_t nrow,
+                                            uint32_t ntrain);
+
+  template <typename value_t>
+  cuivflStatus_t cuivflSearchImpl(
+    const T* queries, uint32_t batch_size, uint32_t k, size_t* neighbors, value_t* distances);
+};
+
+template <typename T>
+cuivflHandle<T>::cuivflHandle(const handle_t& handle,
+                              raft::distance::DistanceType metric_type,
+                              uint32_t dim,
+                              uint32_t nlist,
+                              uint32_t niter,
+                              uint32_t device)
+  : handle_(handle),
+    stream_(handle_.get_stream()),
+    dim_(dim),
+    nlist_(nlist),
+    niter_(niter),
+    metric_type_(metric_type),
+    grid_dim_x_(0),
+    list_data_dev_(0, stream_),
+    list_index_dev_(0, stream_),
+    list_prefix_interleaved_dev_(0, stream_),
+    list_lengths_dev_(0, stream_),
+    centriod_dev_(0, stream_),
+    centriod_norm_dev_(0, stream_),
+    list_index_host_(0),
+    list_prefix_interleaved_host_(0),
+    list_lengths_host_(0),
+    list_data_host_(0)
+{
+  veclen_ = 16 / sizeof(T);
+  while (dim % veclen_ != 0) {
+    veclen_ = veclen_ >> 1;
+  }
+}
```

Comment:
Since we copy the dataset to create an interleaved layout, the `dim % veclen == 0` restriction could be removed by padding the dataset. Please add a todo note. Please also create an issue (if we do not have already) to collect such items, and refer that issue here.